akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,773 @@
+
+    val consumerTracking: ConcurrentHashMap[String, ConsumerTracking] = new ConcurrentHashMap
+    @volatile private var firehoseIsShutdown = false
+
+    private def entityType = firehoseKey.entityType
+    private val sliceRangeStr = s"${firehoseKey.sliceRange.min}-${firehoseKey.sliceRange.max}"
+
+    private def consumerTrackingValues(): Vector[ConsumerTracking] = {
+      import akka.util.ccompat.JavaConverters._
+      consumerTracking.values.iterator.asScala.filter(h => h.history.nonEmpty && h.firehoseOnly).toVector
+    }
+
+    def consumerStarted(consumerId: String, consumerKillSwitch: KillSwitch): Unit = {
+      log.debug("Firehose entityType [{}] sliceRange [{}] consumer [{}] started", entityType, sliceRangeStr, consumerId)
+      consumerTracking.putIfAbsent(
+        consumerId,
+        ConsumerTracking(consumerId, history = Vector.empty, firehoseOnly = false, consumerKillSwitch, None))
+    }
+
+    def consumerTerminated(consumerId: String): Int = {
+      log.debug(
+        "Firehose entityType [{}] sliceRange [{}] consumer [{}] terminated",
+        entityType,
+        sliceRangeStr,
+        consumerId)
+      consumerTracking.remove(consumerId)
+      consumerTracking.size
+    }
+
+    def shutdownFirehoseIfNoConsumers(): Boolean = {
+      if (consumerTracking.isEmpty) {
+        log.debug("Firehose entityType [{}] sliceRange [{}] is shutting down, no consumers", entityType, sliceRangeStr)
+        firehoseIsShutdown = true
+        firehoseKillSwitch.shutdown()
+        true
+      } else
+        false
+    }
+
+    def isShutdown: Boolean =
+      firehoseIsShutdown
+
+    @tailrec def updateConsumerTracking(
+        consumerId: String,
+        now: Instant,
+        offset: TimestampOffset,
+        consumerKillSwitch: KillSwitch): Unit = {
+
+      val existingTracking = consumerTracking.get(consumerId)
+      existingTracking match {
+        case null =>
+          val newTracking =
+            ConsumerTracking(consumerId, history = Vector(offset), firehoseOnly = false, consumerKillSwitch, None)
+          if (consumerTracking.putIfAbsent(consumerId, newTracking) == null)
+            logUpdateConsumerTracking(consumerId, now, newTracking)
+          else
+            // concurrent update, try again
+            updateConsumerTracking(consumerId, now, offset, consumerKillSwitch)
+
+        case existing =>
+          val newHistory =
+            if (existing.history.size <= settings.broadcastBufferSize)
+              existing.history :+ offset
+            else
+              existing.history.tail :+ offset // drop one, add one
+          val newTracking = existing.copy(history = newHistory)
+          if (consumerTracking.replace(consumerId, existingTracking, newTracking))
+            logUpdateConsumerTracking(consumerId, now, newTracking)
+          else
+            // concurrent update, try again
+            updateConsumerTracking(consumerId, now, offset, consumerKillSwitch)
+      }
```

Comment:
This is fine, but just curious if we couldn't use `existingTracking.update(key, (existing) => if (existing == null) ... else ...)` instead for some reason I don't see, is it to avoid the lambda alloc?